src/simulator/simulator.c

```diff
@@ -1,95 +1,78 @@
 #include "simulator/simulator.h"
 
 
-#define SIMULATOR_INITIAL_NUMBER_OF_CALLBACKS 3
-
-
 internal Simulator*
-simulator_create(Network* network, DataGen* data) {
+simulator_create(State* state, Network* network, DataGen* data) {
     Simulator* simulator = NULL;
+    
+    check(state != NULL, "state is NULL");
     check(network != NULL, "network is NULL");
     check(data != NULL, "data is NULL");
     
-    simulator = (Simulator*) memory_calloc(1, sizeof(*simulator), "simulator_create");
+    simulator = (Simulator*) memory_arena_push(state->permanent_storage, sizeof(*simulator));
     check_memory(simulator);
     
     simulator->network = network;
     simulator->data = data;
+    simulator->callback_list = NULL;
     
-    simulator->n_callbacks = 0;
-    simulator->n_max_callbacks = SIMULATOR_INITIAL_NUMBER_OF_CALLBACKS;
-    simulator->callbacks = (CallbackP*) memory_malloc(sizeof(CallbackP) *
-                                                      simulator->n_max_callbacks,
-                                                      "simulator_create callbacks");
-    check_memory(simulator->callbacks);
     return simulator;
     
     error:
-    if (simulator != NULL) {
-        if (simulator->network != NULL) network_destroy(simulator->network);
-        if (simulator->data != NULL) data_gen_destroy(simulator->data);
-        if (simulator->callbacks != NULL) memory_free(simulator->callbacks);
-    }
-    
     return NULL;
 }
 
 
 internal void
-simulator_destroy(Simulator* simulator) {
-    check(simulator != NULL, "simulator is NULL");
-    
-    network_destroy(simulator->network);
-    data_gen_destroy(simulator->data);
-    
-    for (u32 i = 0; i < simulator->n_callbacks; ++i)
-        callback_destroy(simulator->callbacks[i]);
-    memory_free(simulator->callbacks);
-    
-    memset(simulator, 0, sizeof(*simulator));
-    memory_free(simulator);
-    
-    error:
-    return;
-}
-
-
-internal void
-simulator_run(Simulator* simulator)
+simulator_run(State* state, Simulator* simulator)
 {
     TIMING_COUNTER_START(SIMULATOR_RUN);
     
+    check(state != NULL, "state is NULL");
     check(simulator != NULL, "simulator is NULL");
+    
     DataSample* sample = NULL;
     NetworkInputs* inputs = NULL;
-    Callback* callback = NULL;
+    CallbackNode* callback_node = NULL;
     u32 data_idx = 0;
     u32 time = 0;
     u32 i = 0;
     
+    
     for (data_idx = 0; data_idx < simulator->data->length; ++data_idx) {
-        sample = data_gen_sample_create(simulator->data, data_idx);
+        sample = data_gen_sample_create(state, simulator->data, data_idx);
         
         network_clear(simulator->network);
         
-        for (i = 0; i < simulator->n_callbacks; ++i)
-            callback_begin_sample(simulator->callbacks[i], simulator->network, sample->duration);
+        for (callback_node = simulator->callback_list;
```

Comment:
Should i just add an array? Usually I don't have that many callbacks and we shouldn't care here

Reply:
Done